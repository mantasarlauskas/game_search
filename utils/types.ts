import { NextPageContext } from 'next';

export interface Info {
    id: number;
    name: string;
    url?: string;
}

export interface Game {
    name: string;
    slug: string;
    website?: string;
    background_image?: string;
    released?: string;
    platforms?: { platform: Info }[];
    genres: Info[];
    rating: number;
    ratings: { title: string, count: number }[];
    description?: string;
    background_image_additional?: string;
    ratings_count: number;
    metacritic?: number;
    developers: Info[];
    publishers: Info[];
    tags: Info[];
    esrb_rating?: Info;
    clip?: { clip: string };
}

export interface Movie {
    data: {
        480: string,
        max: string,
    }
}

export interface Screenshot {
    image: string;
}

export interface NextPageContextWithID extends NextPageContext {
    params: { id: string }
}

export interface Category {
    id: number;
    name: string;
    slug: string;
    image_background: string;
    games_count: number;
    games: { name: string; slug: string; }[];
}

export interface CategoriesPageProps {
    categories: Category[];
    nextPage?: string;
}

export interface CategoryPageProps {
    id: string;
    games: Game[];
    count: number;
    nextPage?: string;
}

export interface CategoryPageWithNameProps extends CategoryPageProps {
    name: string;
}
