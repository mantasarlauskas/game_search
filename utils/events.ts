import { MouseEvent, KeyboardEvent } from 'react';

export type ElementEvent = MouseEvent | KeyboardEvent;

export function onEnter(fn?: (e: ElementEvent) => void) {
    return (e: ElementEvent) => {
        if ((e as KeyboardEvent).key === 'Enter') {
            fn?.(e);
        }
    };
}
