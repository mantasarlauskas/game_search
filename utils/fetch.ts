export enum API_PATH {
    GAMES = 'games',
}

export async function fetchData(path: string, query?: string) {
    return fetch(
        `https://api.rawg.io/api/${path}?${query || ''}`,
    ).then((res) => {
        if (!res.ok) {
            return null;
        }

        return res.json();
    });
}
