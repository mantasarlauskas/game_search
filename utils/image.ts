export function cropImageUrl(imageUrl?: string) {
    if (!imageUrl) {
        return undefined;
    }

    const gamesIndex = imageUrl.indexOf('/games');
    const index = gamesIndex > 0 ? gamesIndex : imageUrl.indexOf('/screenshots');
    return [imageUrl.slice(0, index), '/crop/600/400', imageUrl.slice(index)].join('');
}
