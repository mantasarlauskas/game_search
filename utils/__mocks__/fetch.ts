import { getGames } from 'testing/testFactories';

module.exports = {
    fetchData: jest.fn().mockResolvedValue({
        results: getGames(),
        count: 100,
        next: '2',
        name: 'name',
    }),
    ApiPath: {
        GAMES: 'games',
        GENRES: 'genres',
        PLATFORMS: 'platforms',
        TAGS: 'tags',
        DEVELOPERS: 'developers',
        PUBLISHERS: 'publishers',
        STORES: 'stores',
    },
};
