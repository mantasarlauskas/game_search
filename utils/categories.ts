import { ApiPath, fetchData, QueryParams } from 'utils/fetch';
import { Route } from 'utils/routes';
import { Category } from 'utils/types';
import { DEFAULT_PAGE_SIZE } from 'utils/page';
import { ParsedUrlQuery } from 'querystring';
import { getOrdering } from 'utils/ordering';

export function getCategoryName(categories: Category[], id: string) {
    return categories.find((category) => category.id === parseInt(id, 10))
        ?.name;
}

export async function getCategoryPageServerSideProps(
    query: ParsedUrlQuery,
    id: string,
    queryParams: QueryParams
) {
    const data = await fetchData(ApiPath.GAMES, {
        page: 1,
        page_size: DEFAULT_PAGE_SIZE,
        ordering: getOrdering(query),
        ...queryParams,
    });

    if (!data?.count || !data?.results) {
        return {
            redirect: {
                destination: Route.NOT_FOUND,
                permanent: false,
            },
        };
    }

    return {
        props: {
            id,
            count: data.count,
            games: data.results,
            nextPage: data.next || null,
        },
    };
}

export async function getCategoryAndGamesServerSideProps(
    query: ParsedUrlQuery,
    path: ApiPath,
    id: string,
    queryParams: QueryParams
) {
    const [data, category] = await Promise.all([
        fetchData(ApiPath.GAMES, {
            page: 1,
            page_size: DEFAULT_PAGE_SIZE,
            ordering: getOrdering(query),
            ...queryParams,
        }),
        fetchData(`${path}/${id}`),
    ]);

    if (!data?.count || !data?.results || !category?.name) {
        return {
            redirect: {
                destination: Route.NOT_FOUND,
                permanent: false,
            },
        };
    }

    return {
        props: {
            id,
            count: data.count,
            games: data.results,
            name: category.name,
            nextPage: data.next || null,
        },
    };
}

export async function getCategoriesPageServerSideProps(path: ApiPath) {
    const data = await fetchData(path, {
        page: 1,
        page_size: DEFAULT_PAGE_SIZE,
    });

    return {
        props: {
            categories: data?.results || [],
            nextPage: data?.next || null,
        },
    };
}
