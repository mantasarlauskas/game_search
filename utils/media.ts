import { Movie } from 'utils/types';

export function getCroppedImageUrl(imageUrl?: string) {
    if (!imageUrl) {
        return undefined;
    }

    const gamesIndex = imageUrl.indexOf('/games');
    const index = gamesIndex > 0 ? gamesIndex : imageUrl.indexOf('/screenshots');
    return [imageUrl.slice(0, index), '/crop/600/400', imageUrl.slice(index)].join('');
}

export function getMovieUrl(movies: Movie[]) {
    return movies[Math.floor(Math.random() * movies.length)]?.data?.[480] || null;
}
