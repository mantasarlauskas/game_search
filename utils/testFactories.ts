import { Category } from 'utils/types';

export function getCategories(): Category[] {
    return [
        {
            id: 18893,
            name: 'Feral Interactive',
            slug: 'feral-interactive',
            games_count: 107,
            image_background: 'https://media.rawg.io/media/games/157/15742f2f67eacff546738e1ab5c19d20.jpg',
            games: [
                {
                    slug: 'rise-of-the-tomb-raider',
                    name: 'Rise of the Tomb Raider',
                },
                {
                    slug: 'shadow-of-mordor',
                    name: 'Middle-earth: Shadow of Mordor',
                },
                {
                    slug: 'hitman',
                    name: 'Hitman',
                },
                {
                    slug: 'bioshock-2',
                    name: 'BioShock 2',
                },
                {
                    slug: 'bioshock-remastered',
                    name: 'BioShock Remastered',
                },
                {
                    slug: 'xcom-enemy-unknown',
                    name: 'XCOM: Enemy Unknown',
                },
            ],
        },
        {
            id: 405,
            name: 'Ubisoft',
            slug: 'ubisoft',
            games_count: 297,
            image_background: 'https://media.rawg.io/media/games/5e4/5e4bff02098b2b6f824c68286d5da1a6.jpg',
            games: [
                {
                    slug: 'far-cry-3',
                    name: 'Far Cry 3',
                },
                {
                    slug: 'for-honor',
                    name: 'For Honor',
                },
                {
                    slug: 'watch-dogs',
                    name: 'Watch Dogs',
                },
                {
                    slug: 'far-cry-5',
                    name: 'Far Cry 5',
                },
                {
                    slug: 'assassins-creed-origins',
                    name: 'Assassin\'s Creed Origins',
                },
                {
                    slug: 'rayman-legends',
                    name: 'Rayman Legends',
                },
            ],
        },
    ];
}
