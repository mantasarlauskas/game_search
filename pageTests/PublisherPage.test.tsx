import { render, screen } from '@testing-library/react';
import PublisherPage, { getServerSideProps } from 'pages/publishers/[id]';
import { getGames } from 'testing/testFactories';
import { fetchData } from 'utils/fetch';
import { NextPageContextWithID } from 'utils/types';

jest.mock('hooks/usePaginatedQuery');
jest.mock('utils/fetch');
jest.mock('next/router', () => ({
    useRouter() {
        return {
            query: '',
        };
    },
}));

describe('<PublisherPage />', () => {
    const props = {
        games: getGames(),
        count: 100,
        id: '123',
        name: 'Ubisoft',
        nextPage: '2',
    };

    it('matches snapshot', () => {
        const { container } = render(<PublisherPage {...props} />);
        expect(container).toMatchSnapshot();
    });

    it('shows content', () => {
        render(<PublisherPage {...props} />);
        expect(screen.getByText('Published by Ubisoft')).toBeInTheDocument();
        expect(screen.getByText('Tomb Raider (2013)')).toBeInTheDocument();
        expect(screen.getByText('Portal 2')).toBeInTheDocument();
        expect(screen.getByText('Total 100 games')).toBeInTheDocument();
    });

    describe('getServerSideProps', () => {
        it('returns server side props', async () => {
            const res = await getServerSideProps({
                params: { id: '123' },
            } as NextPageContextWithID);

            expect(res).toEqual({
                props: {
                    id: '123',
                    count: 100,
                    name: 'name',
                    games: getGames(),
                    nextPage: '2',
                },
            });

            expect(fetchData).toBeCalledTimes(2);
            expect(fetchData).toBeCalledWith('publishers/123');
            expect(fetchData).toBeCalledWith(
                'games',
                expect.objectContaining({
                    publishers: '123',
                })
            );
        });
    });
});
