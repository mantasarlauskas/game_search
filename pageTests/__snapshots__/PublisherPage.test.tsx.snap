// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PublisherPage /> matches snapshot 1`] = `
<div>
  <div
    class="root"
  >
    <h1
      class="title"
    >
      Published by Ubisoft
    </h1>
    <div
      class="header"
    >
      <button
        class="root"
        type="button"
      >
        Order by:
        <div
          class="title"
        >
          Relevance
        </div>
      </button>
      <div
        class="count"
      >
        Total 100 games
      </div>
    </div>
    <div
      class="cards"
    >
      <div>
        <div
          class="root"
        >
          <div
            class="media"
            role="button"
            style="height: auto;"
            tabindex="0"
          >
            <div>
              <img
                alt="Portal 2"
                class="image"
                height="400"
                layout="responsive"
                quality="50"
                src="https://media.rawg.io/media/crop/600/400/games/328/3283617cb7d75d67257fc58339188742.jpg"
                width="600"
              />
            </div>
          </div>
          <div
            class="content"
          >
            <div
              class="header"
            >
              <div
                class="title"
              >
                Portal 2
              </div>
              <div
                class="score"
              >
                4.6
              </div>
            </div>
            <div
              class="platforms"
            >
              Platforms: 
              <div>
                <span
                  class="item"
                >
                  PC, 
                </span>
              </div>
              <div>
                <span
                  class="item"
                >
                  Xbox One
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          class="root"
        >
          <div
            class="media"
            role="button"
            style="height: auto;"
            tabindex="0"
          >
            <div>
              <img
                alt="Tomb Raider (2013)"
                class="image"
                height="400"
                layout="responsive"
                quality="50"
                src="https://media.rawg.io/media/crop/600/400/games/ad2/ad2ffdf80ba993654f31da045bc02456.jpg"
                width="600"
              />
            </div>
          </div>
          <div
            class="content"
          >
            <div
              class="header"
            >
              <div
                class="title"
              >
                Tomb Raider (2013)
              </div>
              <div
                class="score"
              >
                4.1
              </div>
            </div>
            <div
              class="platforms"
            >
              Platforms: 
              <div>
                <span
                  class="item"
                >
                  PC, 
                </span>
              </div>
              <div>
                <span
                  class="item"
                >
                  Xbox One
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="button"
    >
      <button
        class="root"
        type="button"
      >
        Load more
      </button>
      <div
        class="root hidden"
        role="button"
        tabindex="0"
      >
        Back to top
      </div>
    </div>
  </div>
</div>
`;
